lsat/client_interceptor.go

```diff
@@ -249,6 +256,29 @@ func (i *ClientInterceptor) handlePayment(iCtx *interceptContext) error {
 		log.Infof("Payment of LSAT token is required, resuming/" +
 			"tracking previous payment from pending LSAT token")
 		err := i.trackPayment(iCtx.mainCtx, iCtx.token)
+
+		// If the payment failed for good, it will never come back to a
+		// success state. We need to remove the pending token and try
+		// again.
+		if err == errPaymentFailedTerminally {
+			iCtx.token = nil
+			if err := i.store.RemovePendingToken(); err != nil {
+				return fmt.Errorf("error removing pending "+
+					"token, cannot retry payment: %v", err)
+			}
+
+			// Let's try again by paying for the new token.
+			log.Infof("Retrying payment of LSAT token invoice")
+			var err error
+			iCtx.token, err = i.payLsatToken(
```

Comment:
Will this result in us continually retrying the payment if say, the user has no channels or one connected to a very distant edge?

Reply:
It seems that per RPC invocation, if there isn't a pending token, a new payment will be initiated for a new token, and if that fails, we'll try again. When a pending token exists however, a new payment for a new token is only attempted once.

Reply:
Almost correct. If there is no token at all, we initiate a new payment. That can only result in two outcomes: it succeeds and we continue, it fails (including timing out) and we error out. So there is no re-try for the case where we initiate the payment because of the timeout case (the payment could still be in-flight).
What changed is what happens if the user tries again: If we now see the payment is in a terminal state, we remove it and go into the "new payment" flow as described above.

tl;dr: there won't be an automatic re-try loop. but the user can manually try the same call again without needing to delete the token until eventually a payment succeeds.